colossalai/kernel/cuda_native/flash_attention.py

```diff
@@ -0,0 +1,326 @@
+"""
+Fused Attention
+===============
+This is a Triton implementation of the Flash Attention algorithm
+(see: Dao et al., https://arxiv.org/pdf/2205.14135v2.pdf; Rabe and Staats https://arxiv.org/pdf/2112.05682v2.pdf; Triton https://github.com/openai/triton)
+"""
+
+import torch
+from subprocess import PIPE, run
+
+import triton
+import triton.language as tl
+
+from flash_attn.flash_attn_interface import flash_attn_unpadded_func
+
+
+def triton_check():
+    def out(command):
+        result = run(command, stdout=PIPE, stderr=PIPE, universal_newlines=True, shell=False)
+        return result.stdout
+    cuda_version = out("nvcc --version")
+    cuda_version = cuda_version.split('release ')[1]
+    cuda_version = cuda_version.split(',')[0]
+    cuda_version = cuda_version.split('.')
+    if len(cuda_version) < 2 or \
+        (int(cuda_version[0]) == 11 and int(cuda_version[1]) >= 4) or \
+        int(cuda_version[0]) > 11:
+        return False
+    return False
+
+TRITON_AVALIABLE = triton_check()
+
+
+@triton.jit
+def _fwd_kernel(
+    Q, K, V, sm_scale,
+    TMP, L, M,  # NOTE: TMP is a scratchpad buffer to workaround a compiler bug
```

Comment:
can you add type hints for each parameter?
for example
TMP : torch.Tensor, L : int

Reply:
Not sure about some param. Will update it with alphafold version flash attention

Reply:
OK